### PR TITLE
Support testcase properties

### PR DESCRIPTION
**Pull request by omus**
Adds support for defining testcase properties. Both #103 and #104 are precursors to this work.

Overview of the work done here:

- Removes `recordproperty` in favor of `record_testset_property`. I believe `recordproperty` was inspired from the pytest fixture [`record_property`](https://docs.pytest.org/en/6.2.x/usage.html#record-property) which is supposed to add properties to testcases.
- Adds `record_test_property` to allow associating properties against Julia test results. I primarily require this functionality to [support this tool](https://docs.ketryx.com/manuals/man-06-test-management#id-3.4.5.-junit-xml). The JUnit XML schema doesn't officially support properties on testcases but [other JUnit reporting tools already support testcase properties](https://github.com/testmoapp/junitxml?tab=readme-ov-file#properties-for-suites-and-cases). Test properties defined on testsets which contain no dependent tests are ignored.

### CODE REVIEW

File: src/deprecated.jl [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    depwarn("`recordproperty` is deprecated, use `record_testset_property` instead.", :recordproperty)
+    record_testset_property(name, val)
+    return nothing
+end
+
+function properties(ts::AbstractTestSet)
+    depwarn("`$properties` is deprecated, use `$testset_properties` instead.", :properties)
+    p = testset_properties(ts)
+    return isnothing(p) ? nothing : Dict(p)
+end
```
Comment (omus) [inline src/deprecated.jl:15]: Probably want to make this PR breaking and avoid the deprecations as we no longer allow properties to be overwritten

Comment (oxinabox) [reply src/deprecated.jl:15]: Was that a feature before though, or a bug?

Comment (oxinabox) [reply src/deprecated.jl:15]: I guess it was a feature, as it was a documented thing you were warned against doing but were allowed to do.
So it probably should indeed be a breaking change.

When we merge this shall we take the jump straight to 1.0.0?
I do not like having a 0.x.y package

**Comment by oxinabox** [comment]
It is unclear to me from the documentation changed in this PR what the actual difference between a `testset_property` and ` test_property` is.

Without understanding that it is hard to review

**Comment by omus** [comment]
> It is unclear to me from the documentation changed in this PR what the actual difference between a `testset_property` and ` test_property` is.
> 
> Without understanding that it is hard to review

I've added some additional details to those docstrings in terms of how they get translated into the JUnit XML report. Hopefully, that addressed your concern. Please feel free to ask additional questions if you're still having a hard time reviewing.

File: src/deprecated.jl (hunk 1 above)
Comment (omus) [reply src/deprecated.jl]: I'm good with jumping to 1.0.0

File: src/testsets.jl [hunk 2]
```diff
@@ -141,6 +146,11 @@ end
 
 Get the properties associated with tests within a testset. Can be extended for custom
 testsets. Will return `nothing` for testsets which do not support test properties.
+
+When generating a JUnit XML report these will be the properties associated with all
+`testcase` elements contained within a `testsuite` element.
```
Comment (oxinabox) [inline src/testsets.jl:151]: So they will be duplicated?

Comment (omus) [reply src/testsets.jl:151]: No. The properties created through `record_testset_property` are associated to the testset containing the call and all testsets nested within that testset. The properties created through `record_test_property` are associated with the tests contained within the testset containing the function call including all nested tests. When rendering the JUnit report this means that calls to `record_testset_property` are associated with `<testsuite>` elements while calls to `record_test_property` are associated with `<testcase>` elements.

For example:
```julia
using Test, TestReports, EzXML

ts = @testset ReportingTestSet "Root" begin
    record_testset_property("foo", 1)
    record_test_property("bar", 2)

    @testset "Inner" begin
        @test 1 == 1
    end

    @test 2 != 1
end;

prettyprint(report(ts))
```
Results in:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="2" failures="0" errors="0">
  <testsuite name="Root" tests="1" failures="0" errors="0" time="0.080" timestamp="2024-05-02T09:00:58.607" hostname="awro" id="0">
    <testcase name="2 != 1" id="1" classname="Root" time="0.000">
      <properties>
        <property name="bar" value="2"/>
      </properties>
    </testcase>
    <properties>
      <property name="foo" value="1"/>
    </properties>
  </testsuite>
  <testsuite name="Root/Inner" tests="1" failures="0" errors="0" time="0.053" timestamp="2024-05-02T09:00:58.634" hostname="awro" id="1">
    <testcase name="1 == 1" id="1" classname="Root/Inner" time="0.011">
      <properties>
        <property name="bar" value="2"/>
      </properties>
    </testcase>
    <properties>
      <property name="foo" value="1"/>
    </properties>
  </testsuite>
</testsuites>
```

**Comment by omus** [comment]
Rebased this PR since the code in #110 was extracted from this PR. Hopefully, this PR is easier to read now. The `test/properties.jl` file unfortunately doesn't have a proper diff due to the file rename and additional modifications

File: src/testsets.jl (hunk 2 above)
Comment (oxinabox) [reply src/testsets.jl:151]: Got it. this clarified things a lot.
Can this be added into the docs somewhere?
Either in the docstring or in the manual?

**Review by oxinabox — APPROVED** [review]
Ok, cool now i understand it.
And this looks good to me.
Thank you for breaking this up.

Just one small comment.
Then once this is in and the other one we will do a 1.0 release

File: src/testsets.jl (hunk 2 above)
Comment (omus) [reply src/testsets.jl:151]: I've updated the examples in the docstring for `record_testset_properties` to include some of these details. I ended up not including the raw XML output as I thought it was getting a bit too verbose. Hopefully, the example is as clear as this comment. Let me know what you think

**Pull request merged**